### PR TITLE
[SPARK-48835] Introduce versoning to jdbc connectors

### PR DESCRIPTION
**Pull request by milastdbx**
### What changes were proposed in this pull request?

In this PR I propose that we introduce versioning of connectors that is different of spark versioning.
Each connector should have its own version. This way we decouple spark version from connector versions, and customers can upgrade spark without the need to modify connectors as well.
Often times we introduce breaking changes (such as type mapping for example, which can cause schema do not match exception for hive tables).

I suggest we introduce 2 level of versions that each connector will have:
 - JDBC version (for common changes, such as `JdbcUtils`)
 - Dialect version (for dialect specific changes, such as `PostgresDialect`)

Idea is to always roll forward and avoid config explosion.

### Why are the changes needed?

To decouple spark version from connector version.

### Does this PR introduce _any_ user-facing change?

Exposes new configs.

### How was this patch tested?

This is config change.

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala [hunk 1]
```diff
@@ -920,4 +931,15 @@ object JdbcDialects {
  */
 private[spark] object NoopDialect extends JdbcDialect {
   override def canHandle(url : String): Boolean = true
+
+  /**
+   * Noop dialect always uses latest version
+   */
+  override def jdbcVersion(): Int = Int.MaxValue
```
Comment (urosstan-db) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala:938]: No op dialect can be used for some connectors that are not specialized (idk maybe Duck, SAP, Terraform), if we change something in base JDBC dialect, do we need versioning as well there?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 2]
```diff
@@ -4319,6 +4319,146 @@ object SQLConf {
...
+    .version("4.0.0")
+    .intConf
+    .createWithDefault(0)
+
+  val TERADATA_DIALECT_VERSION = buildConf("spark.sql.connector.teradata.dialectVersion")
+    .internal()
+    .doc("Dialect version used for teradata connector.")
+    .version("4.0.0")
+    .intConf
+    .createWithDefault(0)
```
Comment (urosstan-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:4460]: Do we need major and minor versions here, maybe string value is better for versioning

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/DerbyDialect.scala [hunk 3]
```diff
@@ -22,11 +22,22 @@ import java.util.Locale
...
+import org.apache.spark.sql.internal.SQLConf
 import org.apache.spark.sql.types._
 
 
 private case class DerbyDialect() extends JdbcDialect with NoLegacyJDBCError {
 
+  /**
+   * Returns jdbc version that this dialect uses.
+   */
+  override def jdbcVersion(): Int = SQLConf.get.getConf(SQLConf.DERBY_JDBC_VERSION)
```
Comment (urosstan-db) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/DerbyDialect.scala:34]: Do we need jdbcVersion per dialect?

**Review by andrej-db — APPROVED** [review]

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/DerbyDialect.scala (hunk 3 above)
Comment (milastdbx) [reply sql/core/src/main/scala/org/apache/spark/sql/jdbc/DerbyDialect.scala:34]: There are updates on Jdbc common level.
How would we version connector if this is not per dialect ?

Blast radius of these changes would be much higher.

**Comment by milastdbx** [comment]
> @milastdbx do we need to have a doc / design describing the approach?
I don't think so, this should be intiuitive. If its not then we need to redesign until it is

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala (hunk 1 above)
Comment (milastdbx) [reply sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala:938]: thats a good question. I intentionally set that these defaults always pick up latest version (by setting `Int.MaxValue`) but what you are saying makes sense as well

**Review by urosstan-db — APPROVED** [review]

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/DerbyDialect.scala (hunk 3 above)
Comment (urosstan-db) [reply sql/core/src/main/scala/org/apache/spark/sql/jdbc/DerbyDialect.scala:34]: Yeah, I agree, one config for all connectors for common code would be worse solution.
For example, customer can upgrade MySQL independently from Postgres, even common parts of code.

**Comment by kostabiz-db** [comment]
@milastdbx missing Salesforce dialect

**Comment by milastdbx** [comment]
@yaooqinn can you please provide feedback on this PR ?

**Comment by yaooqinn** [comment]
The use of version-numbered configurations in Spark can be unfriendly for users compared to legacy configurations. Various version policies in Spark, such as Thrift Server Versions, API versions, JDBC Spec versions, and others, are implemented to ensure backward compatibility. However, the current system is not user-friendly. It's also hard for us to write user-friendly documentation like https://spark.apache.org/docs/4.0.0-preview1/sql-data-sources-jdbc.html#data-type-mapping.

The rules for data type mapping cannot be determined by dialect versions alone. Various factors such as upstream jdbc artifacts, database server modes, user-specified connection properties, etc., can influence the rules. I believe it's challenging for users to guarantee consistent behavior simply by setting the old version they use.

Given that the data type mapping rules are publicly available in the documentation, see https://spark.apache.org/docs/4.0.0-preview1/sql-data-sources-jdbc.html#data-type-mapping, I'm not convinced that any more radical changes will be made in the future that will require the introduction of versions to control.

**Comment by yaooqinn** [comment]
Also, any number of features in a version can lead to backporting issues.

**Comment by milastdbx** [comment]
@yaooqinn 
> However, the current system is not working in that way.
What do you mean?

> The rules for data type mapping cannot be determined by dialect versions alone. Various factors such as upstream jdbc artifacts, database server modes, user-specified connection properties, etc., can influence the rules. I believe it's challenging for users to guarantee consistent behavior simply by setting the old version they use.

JDBC Artifacts - this one is a good point - we should ideally guard version of driver we use behind config as well. 

For user specified connection properties - I don't understand this point. lets say that we have a flag user-specified connection property A and customer has set it. When this property is set we map javaSQLTypeA to sparkSQLTypeA, but now we want to change this behavior from javaSQLTypeA to sparkSQLTypeB (everything is still under the user-specified property A). We would still need either a flag or version to protect this change.

database server modes  - can you provide an example?

> I'm not convinced that any more radical changes will be made in the future that will require the introduction of versions to control.
Perhaps you are right, but I wouldn't be so certain. With collation support this might change.
Ideally I'd like to achieve two goals with this PR:
 - Avoid flag explosion
 - Encourage people to protect their code with flag (by giving them easy versioning mechanism)

> Also, any number of features in a version can lead to backporting issues.
Isn't this a purpose of a version ?

**Comment by milastdbx** [comment]
@yaooqinn any thoughts on this ?

**Comment by yaooqinn** [comment]
> Isn't this a purpose of a version ? How is version different than a bunch of flags ?

Assuming we have 2 fixes in master, but only the second gets backported to one or more maintenance branches, what values should be set to all these versions in different branches?

**Pull request closed without merging**